File: leetcode/647.cpp
#include "leetcode.h"
/*
    提交前需要把所有输出删除，否则输出时间也会算入运行时间
    最长回文串的变体
    dp[i,j] (i...j)是否为回文
    dp[0,0] = 1
    dp[i,i] = 1
    dp[i,j] = 0 if i>j
    dp[i,j] = dp[i+1,j-1]&&s[i]==s[j]
 */
class Solution
{
public:
    int countSubstrings(string s)
    {
        int len = s.length();
        int dp[len][len];
        memset(dp, 0, sizeof(int) * len * len);
        int result = 0;
        for (int i = 0; i < len - 1; i++)
        {
            dp[i][i] = 1;
            dp[i][i + 1] = (s[i] == s[i + 1]);
        }
        dp[len - 1][len - 1] = 1;
        for (int d = 2; d < len; d++)
        {
            for (int i = 0; i < (len - d); i++)
            {
                int j = i + d;
                dp[i][j] = dp[i + 1][j - 1] && (s[i] == s[j]);
            }
        }
        for (int i = 0; i < len; i++)
        {
            for (int j = i; j < len; j++)
            {
                result += dp[i][j];
            }
        }
        return result;
    }
};
int main()
{
    Solution sol;
    cout << sol.countSubstrings("aaa") << endl;
}

File: leetcode/647.py
class Solution(object):
    def countSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        dp = [[False for i in range(0, len(s))] for i in range(len(s))]
        result = 0
        for i in range(0, len(s)-1):
            dp[i][i], result = True, result+1
            dp[i][i+1] = (s[i] == s[i+1])
            if dp[i][i+1]:
                result += 1
        dp[len(s)-1][len(s)-1], result = True, result+1
        for d in range(2, len(s)):
            for i in range(0, len(s) - d):
                j = i + d
                dp[i][j] = (dp[i+1][j-1]) and (s[i] == s[j])
                if dp[i][j]:
                    result += 1
        return result


'''
dp[i,j]: s[i,j] 是否为回文串
dp[i,j] = invalid, if i > j
dp[i,i] = 1
dp[i,i+1] = (s[i] == s[i+1])
dp[i,j] = dp[i+1,j-1] && (s[i] == s[j])
'''

so = Solution()
print(so.countSubstrings('aaa'))
